GitHubBranches.ps1

```diff
@@ -154,6 +155,550 @@ filter Get-GitHubRepositoryBranch
+
+        Sets a branch protection rule for the master branch of the PowerShellForGithub repository
+        enforcing all configuration restrictions for administrators.
+
+    .EXAMPLE
+        Set-GitHubRepositoryBranchProtectionRule -Uri 'https://github.com/microsoft/PowerShellForGitHub' -BranchName master -RequiredApprovingReviewCount 1
+
+        Sets a branch protection rule for the master branch of the PowerShellForGithub repository
+        requiring one approving review.
+#>
+    [CmdletBinding(
+        PositionalBinding = $false,
+        SupportsShouldProcess,
+        DefaultParameterSetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSReviewUnusedParameter', '',
+        Justification='One or more parameters (like NoStatus) are only referenced by helper methods which get access to it from the stack via Get-Variable -Scope 1.')]
+    param(
+        [Parameter(
+            Mandatory,
+            ValueFromPipelineByPropertyName,
+            Position = 1,
+            ParameterSetName='Uri')]
+        [string] $Uri,
+
+        [Parameter(
+            Mandatory,
+            ValueFromPipelineByPropertyName,
+            Position = 2)]
+        [string] $BranchName,
+
+        [Parameter(ParameterSetName='Elements')]
+        [string] $OwnerName,
+
+        [Parameter(ParameterSetName='Elements')]
+        [string] $RepositoryName,
+
+        [string[]] $StatusChecks,
+
+        [switch] $RequireUpToDateBranches,
+
+        [switch] $EnforceAdmins,
+
+        [string[]] $DismissalUsers,
+
+        [string[]] $DismissalTeams,
+
+        [switch] $DismissStaleReviews,
+
+        [switch] $RequireCodeOwnerReviews,
+
+        [ValidateRange(1, 6)]
+        [int] $RequiredApprovingReviewCount,
+
+        [string[]] $RestrictPushUsers,
+
+        [string[]] $RestrictPushTeams,
+
+        [string[]] $RestrictPushApps,
+
+        [switch] $RequireLinearHistory,
+
+        [switch] $AllowForcePushes,
+
+        [switch] $AllowDeletions,
+
+        [string] $AccessToken,
+
+        [switch] $NoStatus
+    )
+
+    $elements = Resolve-RepositoryElements
+    $OwnerName = $elements.ownerName
+    $RepositoryName = $elements.repositoryName
+
+    $telemetryProperties = @{
+        OwnerName = (Get-PiiSafeString -PlainText $OwnerName)
+        RepositoryName = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    if ($PSBoundParameters.ContainsKey('StatusChecks'))
+    {
+        $requiredStatusChecks = @{
+            strict = $RequireUpToDateBranches.ToBool()
+            contexts = $StatusChecks
+        }
+    }
+    else
+    {
+        $requiredStatusChecks = $null
+    }
+
+    $dismissalRestrictions = @{}
+
+    if ($PSBoundParameters.ContainsKey('DismissalUsers'))
+    {
+       $dismissalRestrictions['users'] = $DismissalUsers
+    }
+    if ($PSBoundParameters.ContainsKey('DismissalTeams'))
+    {
+        $teams = Get-GitHubTeam -OwnerName $OwnerName -RepositoryName $RepositoryName |
```

Comment:
Just some insight -- I was writing a `ConverTo-GitHubSlug` filter (in order to be able to avoid this lookup), but had to abandon it.  I did some edge case testing, and determined that the slug name is non-deterministic from the team name.  I created two teams with foreign characters like so:

`國語` - GitHub made this slug `team`
` 富士山` - GitHub made this slug `team-1`, only because it had already made `team`.

Reply:
Done